# Проект: Определение похожих товаров с использованием FAISS и оценка качества рекомендаций

## Описание проекта:

### Цель проекта

Создать систему, способную идентифицировать и рекомендовать пять товаров, наиболее схожих с каждым элементом из проверочного датасета (validation.csv), используя для этого основную базу данных товаров (base.csv). Эффективность и точность предлагаемых рекомендаций необходимо оценить по метрике accuracy@5.

### Используемые технологии

* FAISS для быстрого поиска по схожести в больших наборах данных.
* Scikit-learn для обучения моделей и оценки качества рекомендаций.

### План работы

* Шаг 1: Подготовка данных и анализ
  * Загрузка и первичный анализ данных из base.csv, validation.csv и validation_answer.csv.
  * Преобразование идентификаторов Id и Target из строк в числовые значения для упрощения обработки.
  * Разделение признаков и целевых переменных, нормализация данных.
* Шаг 2: Работа с FAISS
  * Создание и конфигурация GPU-оптимизированного индекса в FAISS для базового набора данных.
  * Поиск топ-5 похожих товаров для каждого элемента в валидационном наборе.
* Шаг 3: Подготовка обучающего датасета
  * Формирование обучающего датасета на основе результатов поиска с использованием FAISS, включая создание признаков, отражающих схожесть между товарами.
  * Определение целевой переменной (Target) на основе правильных ответов из validation_answer.csv.
* Шаг 4: Обучение модей
  * Оценка качества модели с использованием метрики accuracy@5.
* Шаг 5: Анализ результатов
  * Детальный анализ результатов предсказания модели.
  * Сравнение эффективности рекомендаций FAISS и точности классификации модели логистической регрессии.

## Подготовка данных

### Загрузка библиотек, таблиц, инициализация функций и констант

In [1]:
%pip install faiss_cpu

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: python3 -m pip install --upgrade pip


In [2]:
# Базовые библиотеки для работы с данными
import pandas as pd
import numpy as np
import faiss

# Визуализация
import seaborn as sns
import matplotlib.pyplot as plt

# Модели и инструменты машинного обучения
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.metrics import make_scorer, accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression, LogisticRegression

# Утилиты
from IPython.display import display
import warnings
import time


# Игнорирование предупреждений
warnings.filterwarnings('ignore')

In [3]:
pd.set_option('display.min_rows', 15)
pd.set_option('display.max_rows', 20)
pd.options.display.float_format = '{:,.2f}'.format

In [4]:
valid = pd.read_csv('validation.csv')

In [5]:
valid_target= pd.read_csv('validation_answer.csv')

In [6]:
train = pd.read_csv('train.csv')

In [7]:
base = pd.read_csv('base.csv')

In [8]:
# функция информации по таблице
def dataframe_summary(df, string):
    # Вывод общей информации
    print("Общая информация по таблице:", string)
    df.info()

    print("\n Статистическое описание:")
    display(df.describe().transpose())

    print("\nСлучайные примеры:")
    display(df.sample(5))

    print("\nКоличество строк и столбцов:", df.shape)

    print("\nКоличество явных дубликатов:", df.duplicated().sum())
    print('')

In [9]:
def downcast_dataframe(df):
    for col in df.columns:
        if df[col].dtype == 'float64':  # Проверяем, является ли тип столбца float64
            df[col] = pd.to_numeric(df[col], downcast='float')
        elif df[col].dtype == 'int64':  # Проверяем, является ли тип столбца int64
            df[col] = pd.to_numeric(df[col], downcast='integer')
    return df

Все данные загружены, можно приступать к работе.

### Обзор данных

In [10]:
for data in [valid, valid_target, train, base]:
    dataframe_summary(data, 'Данные')

Общая информация по таблице: Данные
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 73 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Id      100000 non-null  object 
 1   0       100000 non-null  float64
 2   1       100000 non-null  float64
 3   2       100000 non-null  float64
 4   3       100000 non-null  float64
 5   4       100000 non-null  float64
 6   5       100000 non-null  float64
 7   6       100000 non-null  float64
 8   7       100000 non-null  float64
 9   8       100000 non-null  float64
 10  9       100000 non-null  float64
 11  10      100000 non-null  float64
 12  11      100000 non-null  float64
 13  12      100000 non-null  float64
 14  13      100000 non-null  float64
 15  14      100000 non-null  float64
 16  15      100000 non-null  float64
 17  16      100000 non-null  float64
 18  17      100000 non-null  float64
 19  18      100000 non-null  float64
 20  19      10000

count    mean    std       min       25%     50%     75%    max
0  100,000.00  -85.30  25.78   -190.35   -103.23  -85.30  -67.42  14.43
1  100,000.00    7.67   4.96    -11.11      4.29    7.66   10.99  27.41
2  100,000.00  -43.84  39.14   -217.54    -69.68  -43.23  -17.77 134.86
3  100,000.00 -146.12  20.50   -220.05   -160.21 -146.08 -132.12 -57.38
4  100,000.00  111.64  47.75    -81.20     79.10  111.96  143.90 302.07
5  100,000.00  -73.27  28.52   -176.71    -92.79  -73.58  -53.86  50.26
6  100,000.00 -441.40 279.24   -791.46   -740.62 -513.92 -202.57 109.63
..        ...     ...    ...       ...       ...     ...     ...    ...
65 100,000.00   14.40  98.70   -157.59    -70.42   14.77   99.80 185.09
66 100,000.00   67.79   1.84     60.38     66.55   67.80   69.04  75.40
67 100,000.00   23.25  55.40   -210.67    -13.86   23.51   60.15 251.29
68 100,000.00   73.11  62.06   -175.92     31.49   72.15  114.32 305.94
69 100,000.00  115.20  21.49     25.27    100.54  115.28  129.77 201.60
70 100,000.00 -709.46 405.67 -1,297.92 -1,074.46 -807.03 -358.40  98.74
71 100,000.00  -48.42  41.29   -209.94    -76.36  -48.67  -20.28 111.83

[72 rows x 8 columns]


Случайные примеры:


Id       0     1       2  ...     68     69        70     71
89980  189980-query -105.72  8.66  -50.96  ... 128.77 102.67   -680.50 -10.56
11200  111200-query  -43.67 10.75  -49.78  ...  22.57  97.24   -929.32 -61.13
4216   104216-query  -97.73  7.20 -104.96  ... -17.29 103.57   -413.29 -83.86
87902  187902-query -116.85 10.21   27.86  ...   8.09  85.90   -345.00 -57.94
10204  110204-query -134.82  3.51    3.32  ...  38.30 130.41 -1,074.46 -19.27

[5 rows x 73 columns]


Количество строк и столбцов: (100000, 73)

Количество явных дубликатов: 0

Общая информация по таблице: Данные
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 2 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   Id        100000 non-null  object
 1   Expected  100000 non-null  object
dtypes: object(2)
memory usage: 1.5+ MB

 Статистическое описание:


count  unique           top freq
Id        100000  100000  100000-query    1
Expected  100000   91502   210304-base    7


Случайные примеры:


Id     Expected
83077  183077-query  331547-base
65511  165511-query  669327-base
62715  162715-query  102401-base
62662  162662-query  121885-base
58720  158720-query  119152-base


Количество строк и столбцов: (100000, 2)

Количество явных дубликатов: 0

Общая информация по таблице: Данные
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 74 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Id      100000 non-null  object 
 1   0       100000 non-null  float64
 2   1       100000 non-null  float64
 3   2       100000 non-null  float64
 4   3       100000 non-null  float64
 5   4       100000 non-null  float64
 6   5       100000 non-null  float64
 7   6       100000 non-null  float64
 8   7       100000 non-null  float64
 9   8       100000 non-null  float64
 10  9       100000 non-null  float64
 11  10      100000 non-null  float64
 12  11      100000 non-null  float64
 13  12      100000 non-null  float64
 14  13      100000 non-null  float64
 15  14      100000 non-null  float64
 16  15      100000 non-null  float64
 17  16      100000 non-null  float64
 18  17      100000

count    mean    std       min       25%     50%     75%    max
0  100,000.00  -85.33  25.80   -186.28   -103.31  -85.27  -67.33  14.59
1  100,000.00    7.66   4.96    -11.56      4.31    7.65   10.98  28.92
2  100,000.00  -43.67  39.11   -224.90    -69.49  -42.83  -17.71 128.11
3  100,000.00 -146.12  20.43   -223.31   -160.03 -146.07 -132.28 -60.75
4  100,000.00  111.77  47.70    -93.27     79.47  112.26  143.76 301.36
5  100,000.00  -73.18  28.72   -184.96    -92.83  -73.36  -53.70  51.85
6  100,000.00 -440.62 279.02   -791.47   -738.54 -511.84 -201.12 109.63
..        ...     ...    ...       ...       ...     ...     ...    ...
65 100,000.00   14.00  99.08   -157.59    -71.75   14.07  100.01 185.10
66 100,000.00   67.80   1.84     60.67     66.56   67.81   69.06  74.78
67 100,000.00   23.03  55.47   -203.75    -14.14   23.44   59.94 266.49
68 100,000.00   73.41  62.20   -181.97     31.69   72.88  114.47 319.87
69 100,000.00  115.19  21.58     22.60    100.59  115.24  129.79 201.76
70 100,000.00 -709.76 405.96 -1,297.87 -1,074.46 -808.80 -357.09  98.77
71 100,000.00  -48.51  41.22   -209.94    -76.48  -48.70  -20.49 126.19

[72 rows x 8 columns]


Случайные примеры:


Id       0     1       2  ...     69        70     71        Target
65198  65198-query  -56.16  8.68  -30.93  ... 122.29 -1,074.46 -90.91   836296-base
83580  83580-query  -85.99 11.83  -67.90  ...  98.83   -386.77 -28.64   503513-base
89714  89714-query -129.28  6.69 -128.10  ...  87.51 -1,074.46 -79.69  3386721-base
13588  13588-query  -76.82  6.93   68.53  ... 130.22   -392.66 -31.55  1201506-base
1394    1394-query  -84.36 -0.91  -30.85  ... 107.90   -213.57 -77.07  1144110-base

[5 rows x 74 columns]


Количество строк и столбцов: (100000, 74)

Количество явных дубликатов: 0

Общая информация по таблице: Данные
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2918139 entries, 0 to 2918138
Data columns (total 73 columns):
 #   Column  Dtype  
---  ------  -----  
 0   Id      object 
 1   0       float64
 2   1       float64
 3   2       float64
 4   3       float64
 5   4       float64
 6   5       float64
 7   6       float64
 8   7       float64
 9   8       float64
 10  9       float64
 11  10      float64
 12  11      float64
 13  12      float64
 14  13      float64
 15  14      float64
 16  15      float64
 17  16      float64
 18  17      float64
 19  18      float64
 20  19      float64
 21  20      float64
 22  21      float64
 23  22      float64
 24  23      float64
 25  24      float64
 26  25      float64
 27  26      float64
 28  27      float64
 29  28      float64
 30  29      float64
 31  30      float64
 32  31      float64
 33  32      float64
 34  33      float6

count    mean    std       min       25%       50%     75%    max
0  2,918,139.00  -86.23  24.89   -199.47   -103.07    -86.23  -69.26  21.52
1  2,918,139.00    8.08   4.95    -13.91      4.71      8.04   11.47  29.94
2  2,918,139.00  -44.58  38.63   -240.07    -69.56    -43.82  -19.63 160.94
3  2,918,139.00 -146.63  19.84   -232.67   -159.91   -146.78 -133.33 -51.37
4  2,918,139.00  111.32  46.35   -105.58     80.51    111.87  142.37 319.66
5  2,918,139.00  -71.99  28.19   -211.01    -91.38    -71.92  -52.44  58.81
6  2,918,139.00 -392.22 271.65   -791.47   -629.33   -422.20 -156.67 109.63
..          ...     ...    ...       ...       ...       ...     ...    ...
65 2,918,139.00   14.15  98.95   -157.59    -71.30     13.83   99.67 185.10
66 2,918,139.00   67.79   1.82     59.51     66.58     67.81   69.03  75.71
67 2,918,139.00   23.54  55.34   -233.14    -12.52     23.42   59.76 314.90
68 2,918,139.00   74.96  61.35   -203.60     33.78     74.93  115.88 339.57
69 2,918,139.00  115.57  21.18     15.72    101.69    116.02  129.55 214.71
70 2,918,139.00 -799.34 385.41 -1,297.93 -1,074.46 -1,074.46 -505.74  98.77
71 2,918,139.00  -47.79  41.75   -226.78    -75.67    -48.59  -19.71 126.97

[72 rows x 8 columns]


Случайные примеры:


Id       0     1      2  ...     68     69        70     71
50380      50950-base -149.02  2.05 -48.13  ...  28.14 148.88 -1,166.76 -25.16
1132499  1403229-base  -88.77  9.33 -45.09  ...  34.25 120.59   -135.92 -67.97
123291    126650-base -111.51 16.00 -53.16  ... 103.73 121.04   -554.62 -59.44
277171    293906-base  -95.27  8.49   3.24  ...  69.39 115.73     38.64 -20.98
1259737  1594335-base -104.06  2.93 -43.18  ... 105.52 149.18 -1,074.46 -30.85

[5 rows x 73 columns]


Количество строк и столбцов: (2918139, 73)

Количество явных дубликатов: 0



* В данных отсутсвуют пропуски и дубликаты. 
* Наблюдается сильный разброс значений, понадобится масштабирование.
* Стоит отменить, что в `valid_target` в колонке `Expected` 91502 уникальных значений из 100000 строк. Значит есть товары, которые указали как схожие несколько раз.

Уменьшим потребляемый объем памяти данных. Переведем столбец Id в формат индекса. И уменьшим регистр называний столбцов.

In [11]:
base = downcast_dataframe(base)
base.columns = base.columns.str.lower()
base.set_index('id', inplace=True)

In [12]:
train = downcast_dataframe(train)
train.columns = train.columns.str.lower()
train.set_index('id', inplace=True)

In [13]:
valid = downcast_dataframe(valid)
valid.columns = valid.columns.str.lower()
valid.set_index('id', inplace=True)

In [14]:
valid_target = downcast_dataframe(valid_target)
valid_target.columns = valid_target.columns.str.lower()
valid_target.set_index('id', inplace=True)

In [15]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100000 entries, 0-query to 99999-query
Data columns (total 73 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   0       100000 non-null  float32
 1   1       100000 non-null  float32
 2   2       100000 non-null  float32
 3   3       100000 non-null  float32
 4   4       100000 non-null  float32
 5   5       100000 non-null  float32
 6   6       100000 non-null  float32
 7   7       100000 non-null  float32
 8   8       100000 non-null  float32
 9   9       100000 non-null  float32
 10  10      100000 non-null  float32
 11  11      100000 non-null  float32
 12  12      100000 non-null  float32
 13  13      100000 non-null  float32
 14  14      100000 non-null  float32
 15  15      100000 non-null  float32
 16  16      100000 non-null  float32
 17  17      100000 non-null  float32
 18  18      100000 non-null  float32
 19  19      100000 non-null  float32
 20  20      100000 non-null  float32
 21  21  

Потребляемый объем памяти упал почти в два раза.

## Поиск ближайших соседей, формирование новой валидационной и тренировочной выборки

#### Алгоритм Faiss

In [16]:
# выделение цел. признака из обучающей выборки
train_target = train['target']
train.drop('target', axis=1, inplace=True)

In [17]:
# словарь id и номеров векторов
base_index = {k: v for k, v in enumerate(base.index.to_list())}

In [18]:
train_index = {k: v for k, v in enumerate(train.index.to_list())}

In [19]:
# Масштабирование 
scaler = StandardScaler()
scaler.fit(base)
base = scaler.transform(base)
train = scaler.transform(train)
valid = scaler.transform(valid)

In [20]:
N_CLUSTERS = 50   # количество кластеров
N_NEIGHBORS = 50   # количество ближайших соседей для поиска

In [21]:
%%time 
# Создаем квантизатор для определения кластеров
d = base.shape[1]  # размерность векторов
quantizer = faiss.IndexFlatL2(d)  

# Создаем индекс для векторов с использованием квантизатора
index = faiss.IndexIVFFlat(quantizer, d, N_CLUSTERS, faiss.METRIC_L2)

# Обучаем индекс на первых 50 000 строках из base
assert not index.is_trained  # убеждаемся, что индекс еще не обучен
index.train(base[:50000])

# Добавляем весь массив base в индекс после обучения
index.add(base)

# Включаем поиск по всему индексу, если не хватает векторов для заполнения кластеров
index.nprobe = 60   

CPU times: user 13.6 s, sys: 622 ms, total: 14.3 s
Wall time: 2.28 s


#### Создание выборки train

In [22]:
# Используем обученный индекс для поиска ближайших соседей в train
D, I = index.search(train, N_NEIGHBORS)

# D - расстояния до ближайших соседей, I - индексы ближайших соседей в массиве base

In [23]:
print(f'train_idx :\n {I[:3]} \n')
print(f'Distance:\n {D[:3]} \n')

train_idx :
 [[ 598613  755584  336969 1934845   13374 1136231  480296  583287 2360257
   450667  988777 1653095 1818641  503716 1631947   89840  143978  986050
   629775 2346335  223859  352715  455829 1747988 1674977  443428 2295048
   728097 1113711 1746258  751217  854264  792610 1375244  199196 2212973
  2398299 1679537  679629  684958  669932  108138 1285249  725525 1546520
   458465 1299775 1307670 2331892  547719]
 [ 737853   15185  519311  902634  223456  464427  220644 2146806 2524968
   643208  134844  301385  340012  298805  216389  188242  971689  573064
  1459914  548516  924284 1010871  917176 1348395   10845 2214631 1312093
   603976  458085  780936  419818  839888  426013 2375011  332349 1096992
   242607 2207064 1946481  838522 1084963 1238219  124343 1182284 1839713
   943287 2865631 2454898   49621 1541125]
 [ 173120 1784963 2863071 1143288  955831  137853 1368891 1162532  774247
  1984148 1479188 1238526 2351134  641280 1532675 1774212  948608   12304
  2395210 190

In [24]:
def accuracy_n(targets, predict, base_index):

    if isinstance(predict, np.ndarray):
        predict = predict.tolist()

    correct_predictions = sum(int(targets[i] in [base_index[idx] for idx in pred]) for i, pred in enumerate(predict))

    accuracy = (correct_predictions / len(predict)) * 100
    return accuracy

In [25]:
accuracy = accuracy_n(train_target, I, base_index)

In [26]:
print(f'Accuracy 50-neighbors: {accuracy:.2f}%')

Accuracy 50-neighbors: 76.76%


Создадим новые таблицы с таргет признаками, которые помогут в дальнейшем анализе.

In [27]:
# подготовка датафрейма с перечнем ID новых и старых объектов
idx_df = pd.DataFrame(data=I, index=train_index)
idx_df = pd.melt(idx_df.T)
idx_df.columns = ['id_query', 'predicted']
idx_df['id_predicted'] = [base_index[number] for number in idx_df['predicted'].values]
idx_df.drop('predicted', axis=1, inplace=True)
idx_df['id_query'] = idx_df['id_query'].map(train_index)


In [28]:
display(idx_df)

id_query  id_predicted
0            0-query   675816-base
1            0-query   877519-base
2            0-query   361564-base
3            0-query  2725256-base
4            0-query    13406-base
5            0-query  1408780-base
6            0-query   530165-base
...              ...           ...
4999993  99999-query  4593791-base
4999994  99999-query  2638659-base
4999995  99999-query   275860-base
4999996  99999-query  2372472-base
4999997  99999-query  1474295-base
4999998  99999-query  1255868-base
4999999  99999-query   396195-base

[5000000 rows x 2 columns]

In [29]:
base_index_df = pd.Series(data=base_index)
train_index_df = pd.Series(data=train_index)

In [30]:
base_transform_df  = pd.DataFrame(base,  index=base_index_df)
train_transform_df = pd.DataFrame(train, index=train_index_df)

In [31]:
base_transform_df.sample(5)

0     1     2     3     4   ...    67    68    69    70    71
532281-base  -0.86  1.48 -1.91 -0.67 -1.47  ... -0.33  0.25  0.59 -0.71  1.14
2470812-base  1.13  0.50  1.78  0.44 -2.31  ... -2.95  0.49 -0.67 -0.71 -1.43
284697-base  -0.99  0.39 -1.03 -0.79 -0.67  ...  0.43 -0.18  0.34  1.12 -0.24
2972932-base  0.22  0.42 -0.62 -1.66  0.73  ... -0.07  0.17 -0.42 -0.71 -1.32
268421-base   1.17 -0.23  1.02  0.73  0.60  ...  0.06 -0.54  0.37 -0.71 -0.90

[5 rows x 72 columns]

In [32]:
train_transform_df.sample(5)

0     1     2     3     4   ...    67   68    69    70    71
72994-query  1.16  1.68 -0.28 -1.11 -0.06  ...  0.04 0.07 -0.54 -0.71  0.63
52956-query  0.54 -0.26 -1.26 -0.44 -1.01  ... -1.49 1.29 -0.55  0.43  0.38
32702-query  1.76  0.58  1.51  2.05  0.66  ...  0.72 2.02  1.11 -0.30 -1.97
44817-query  1.32 -1.14 -0.28  0.29 -0.69  ... -1.36 1.25 -0.60 -0.25 -0.01
46137-query -0.41 -0.97 -0.63 -1.59 -2.78  ...  0.59 0.08  0.82  1.75 -0.21

[5 rows x 72 columns]

In [33]:
train_new = pd.merge(idx_df, train_transform_df, left_on='id_query', right_index=True, how='left')


In [34]:
train_new = pd.merge(train_new, base_transform_df, left_on='id_predicted', right_index=True, how='left')

In [35]:
# Добавляем колонку 'target' в train_new, используя индексы из train_target для сопоставления
train_new['target'] = train_new['id_query'].map(train_target)
train_new['target'] = (train_new['target'] == train_new['id_predicted']).astype(int)
# выделение признаков и таргета
target_train = train_new['target']
features_train = train_new.drop(['id_query', 'id_predicted', 'target'] , axis=1)

Вывод:
Была создана тренировочная выборка на основе поиска 50 ближайших соседей. С точностью в 76%.

#### Создание валидацонной выборки

In [36]:
# Используем обученный индекс для поиска ближайших соседей в train
D, I = index.search(valid, N_NEIGHBORS)

In [37]:
valid_index = {k: v for k, v in enumerate(valid_target.index.to_list())}

In [38]:
valid_target_series = valid_target['expected']

In [39]:
valid_predictions = pd.DataFrame(I)  # I - индексы ближайших соседей
valid_predictions = valid_predictions.stack().reset_index(level=1, drop=True).to_frame('id_predicted')
valid_predictions['id_query'] = valid_predictions.index.map(valid_index)
valid_predictions['id_predicted'] = valid_predictions['id_predicted'].map(base_index)  # Переводим индексы в ID
valid_predictions.set_index('id_query', inplace=True)

In [40]:
merged_valid = valid_predictions.join(valid_target)

In [41]:
# Создаем маску, где true - если expected присутствует среди id_predicted для каждого id_query
mask = merged_valid.groupby(merged_valid.index).apply(lambda x: x['expected'].iloc[0] in x['id_predicted'].values)

# Вычисляем accuracy как процент true значений в маске
accuracy_at_50 = (mask.sum() / len(mask)) * 100
print(f"Точность 50 соседей на валидационной выборке: {accuracy_at_50}%")

Точность 50 соседей на валидационной выборке: 76.696%


In [42]:
# Преобразование матрицы valid в DataFrame
valid_df = pd.DataFrame(valid)
valid_df.index = [valid_index[i] for i in range(valid.shape[0])]

merged_valid_df = merged_valid.join(valid_df)
merged_valid_df = pd.merge(merged_valid_df, base_transform_df, left_on='id_predicted', right_index=True, how='left')

In [43]:
merged_valid_df['expected'] = (merged_valid_df['expected'] == merged_valid_df['id_predicted']).astype(int)

In [44]:
target_merged_valid = merged_valid_df['expected']
features_merged_valid = merged_valid_df.drop(['id_predicted', 'expected'] , axis=1)

In [45]:
del valid, train, base, idx_df, train_new, merged_valid, D, I, mask

Вывод: 

Создали валидационную выборку на основе поиска 50 ближайших соседей.

### Модель Логистической регресии

Обучим модель логистичиской регрессии. Таргет признак - совпадение предсказанного id товара с id товара размеченного. Для каждого товара мы нашли 50 кандидатов и среди них у 76% есть размеченный товар. \
Модель логистической регрессии будет предсказывать случаи, когда id векторов совпадают то класс 1.

In [46]:
# Модель логистической регрессии
model = LogisticRegression(solver='lbfgs', class_weight='balanced', random_state=42)
model.fit(features_train.values, target_train)

LogisticRegression(class_weight='balanced', random_state=42)

In [47]:
valid_predictions = model.predict(features_merged_valid.values)
valid_probabilities = model.predict_proba(features_merged_valid.values)[:, 1]
# Вычисление точности на валидационном наборе
valid_accuracy = accuracy_score(target_merged_valid, valid_predictions)

In [48]:
print('Точность на Валидационной выборке', valid_accuracy)

Точность на Валидационной выборке 0.6010702


In [49]:
valid_proba_series = pd.Series(valid_probabilities)
valid_candidates = []
for i in range(0, len(valid_proba_series), N_NEIGHBORS):
    query = valid_proba_series[i : i+N_NEIGHBORS]
    index = query.sort_values(ascending=False)[0:5].index
    valid_candidates.append(index)

In [50]:
valid_candidates_df = pd.DataFrame(data=valid_candidates)
# Преобразование pd.Series в pd.DataFrame
target_merged_valid_df = target_merged_valid.to_frame()

target_merged_valid_df['prob'] = valid_proba_series.values

target_merged_valid_df['prediction'] = valid_predictions
# Преобразование индекса в столбец
target_merged_valid_df_reset = target_merged_valid_df.reset_index().rename(columns={'index': 'id_query'})

# Группировка по 'id_query' и отбор топ-5 записей с наибольшими вероятностями
top_5_per_group = target_merged_valid_df_reset.groupby('id_query').apply(
    lambda x: x.nlargest(5, 'prob')
)

top_5_per_group = top_5_per_group.reset_index(drop=True)

In [51]:
# Для каждой группы проверяем, есть ли среди топ-5 записей такие, где expected и prediction равны 1
accuracy_per_group = top_5_per_group.groupby('id_query').apply(
    lambda x: ((x['expected'] == 1) & (x['prediction'] == 1)).any()
).mean()

In [52]:
print(f'Accuracy среди топ-5 кандидатов: {accuracy_per_group:.2%}')

Accuracy среди топ-5 кандидатов: 12.81%


Запомним это значение, в будущих обновлнниях попробуем еще модель для сравнения.

## Вывод:


В данном проекте мы проделали следующую работу
* Загрузка и предобработка данных: На этом этапе мы загрузили данные, очистили их от возможных аномалий и пропусков, подготовив таким образом к дальнейшему анализу.

* Масштабирование данных: Следующим шагом было масштабирование признаков для улучшения работы алгоритмов машинного обучения и упрощения процесса обучения модели.

* Применение Faiss для векторного поиска: Использование Faiss позволило нам эффективно разделить векторное пространство на 50 кластеров и найти 50 ближайших соседей для каждого товара из запроса. Это дало возможность ускорить процесс поиска и сделать его более масштабируемым.

* Оценка точности подбора соседей: Используя размеченные данные, мы выяснили, что среди 50 подобранных соседей для каждого товара из запроса у 76% товаров было найдено совпадение по размеченному признаку. Это свидетельствует о высокой точности алгоритма в поиске релевантных товаров.

* Применение модели логистической регрессии: Дальнейшее использование логистической регрессии для выбора 5 наиболее подходящих кандидатов из 50 предварительно отобранных показало долю правильных подборов в 13%. Этот результат указывает на потенциал для улучшения и оптимизации процесса метчинга.

### Возможные направления для улучшений:
Использование более сложных моделей машинного обучения: Вместо логистической регрессии можно применить более сложные и мощные алгоритмы, такие как случайный лес, градиентный бустинг или нейронные сети, которые могут улучшить точность предсказаний.

Расширение набора признаков: Добавление новых признаков, таких как текстовые описания товаров, категории, ценовые диапазоны, может помочь моделям лучше различать товары и повысить точность метчинга.

Оптимизация гиперпараметров: Тонкая настройка гиперпараметров существующих моделей с помощью методов, таких как GridSearchCV или RandomizedSearchCV, может привести к значительному улучшению результатов.

Использование методов глубокого обучения: Применение свёрточных нейронных сетей для анализа изображений товаров или рекуррентных нейронных сетей для обработки их описаний может улучшить способность моделей выделять ключевые характеристики товаров.

In [65]:
del valid_candidates, query, valid_predictions, valid_probabilities, valid_candidates_df, valid_proba_series